### ISSUE TITLE
How to use BlockGrad with weight decay

### ISSUE THREAD

**Issue by VanDavv** [apache/incubator-mxnet#12392]
I saw in [this](https://github.com/apache/incubator-mxnet/issues/1340#issuecomment-174166248)  and [this](https://github.com/apache/incubator-mxnet/issues/3073#issuecomment-241033513) issue comments, @winstywang suggests that using BlockGrad with weight decay is highly inadviseable. As I also asked in [this](https://github.com/apache/incubator-mxnet/issues/3073#issuecomment-416852189) comment, how can I use BlockGrad to freeze first layers of the net, while still using weight decay on the unfrozen ones?

**Comment by anirudhacharya** [comment]
@mxnet-label-bot [Question, How To, Modeling]

**Comment by thomelane** [comment]
Hi @VanDavv,

Given the original context of transfer learning, with Gluon you can just provide the parameters for the last layer to the `Trainer` object. Usually you'd give all of the parameters with `net.collect_params()`, but instead only provide the parameters you're interested in learning/tuning.

See https://mxnet.incubator.apache.org/tutorials/onnx/fine_tuning_gluon.html for an example, and use `dense_layer` if you only want to train the last layer and not fine-tune the whole network:

```
trainer = gluon.Trainer(dense_layer.collect_params(), 'sgd', 
                        {'learning_rate': LEARNING_RATE,
                         'wd':WDECAY,
                         'momentum':MOMENTUM})
```

**Comment by thomelane** [comment]
Answering your question a little more directly (if you haven't moved over to Gluon), you could set the weight decay to zero for the parameters before the BlockGrad with [`set_wd_mult`](https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/optimizer.py#L330).

**Comment by piyushghai** [comment]
@VanDavv Hope your question has been answered :) 
For more questions, please feel free to post on MXNet discussion forum (https://discuss.mxnet.io), where it will get a wider audience and allow other to learn as well.
@sandeep-krishnamurthy Please close this issue. 
@VanDavv Please feel free to re-open if closed in error.
Thanks!"
